orca-core/src/main/java/com/netflix/spinnaker/orca/locks/LockManager.java

```diff
@@ -0,0 +1,121 @@
+package com.netflix.spinnaker.orca.locks;
+
+import java.util.Objects;
+
+/**
+ * Manages acquisition / release of locks.
+ *
+ * The lock is held with a value which is the combination of application, type, and id. This
+ * would typically be the execution application, type (pipeline or orchestration) and execution id.
+ *
+ * A named lock can be acquired if either it is not currently locked, or if the current value of the
+ * lock matches the supplied lockValue fields.
+ *
+ * A lock has a TTL to ensure that in the event of an unexpected failure or JVM exit that eventually locks
+ * are released - the general expectation is that a call to acquireLock should be accompanied by a call to
+ * releaseLock.
+ *
+ * Multiple holders can acquire the same lock by supplying the same lockValue fields, and a lock continues to be
+ * held until all holders have issued a releaseLock.
+ *
+ * A stage that operates on a particular cluster should acquire the lock for that cluster. Any synthetic stages
+ * can also acquire the same lock. For example:
+ *
+ * As an example, a deploy with Red/Black would go through the following:
+ *
+ *
+ * lockValue = new LockValue(application, execution.type, execution.id)
+ * createServerGroup
+ *   acquireLock(clusterName, lockValue, createServerGroup.stage.id)
```

Comment:
do we really need that top-level lock? I'd like to think we would only need it for destructive actions (i.e. the `disableCluster` and `scaleDown` cluster below)